File: №22 Template Pattern/Java/src/Main.java
abstract class MovieBookingTemplate {
    public void bookTicket() {
        selectMovie();
        selectSeats();
        makePayment();
        confirmBooking();
    }

    protected abstract void selectMovie();

    protected abstract void selectSeats();

    protected abstract void makePayment();

    protected void confirmBooking() {
        System.out.println("Booking confirmed!");
    }
}

class MovieBookingProcess extends MovieBookingTemplate {
    private final String movieName;
    private final int numberOfTickets;

    public MovieBookingProcess(String movieName, int numberOfTickets) {
        this.movieName = movieName;
        this.numberOfTickets = numberOfTickets;
    }

    @Override
    protected void selectMovie() {
        System.out.println("Selected Movie: " + movieName);
    }

    @Override
    protected void selectSeats() {
        System.out.println("Selected " + numberOfTickets + " seats.");
    }

    @Override
    protected void makePayment() {
        System.out.println("Payment successful for " + numberOfTickets + " tickets.");
    }
}

public class Main {
    public static void main(String[] args) {

        MovieBookingTemplate bookingProcess = new MovieBookingProcess("Avengers: Endgame", 5);
        bookingProcess.bookTicket();
    }
}

File: №22 Template Pattern/Python/main.py
from abc import ABC, abstractmethod


# Template class
def confirm_booking():
    print("Booking confirmed!")


class MovieBookingTemplate(ABC):
    def book_ticket(self):
        self.select_movie()
        self.select_seats()
        self.make_payment()
        confirm_booking()

    @abstractmethod
    def select_movie(self):
        pass

    @abstractmethod
    def select_seats(self):
        pass

    @abstractmethod
    def make_payment(self):
        pass


class MovieBookingProcess(MovieBookingTemplate):
    def __init__(self, movie_name, number_of_tickets):
        self.movie_name = movie_name
        self.number_of_tickets = number_of_tickets

    def select_movie(self):
        print(f"Selected Movie: {self.movie_name}")

    def select_seats(self):
        print(f"Selected {self.number_of_tickets} seats.")

    def make_payment(self):
        print(f"Payment successful for {self.number_of_tickets} tickets.")


if __name__ == "__main__":
    booking_process = MovieBookingProcess("Avengers: Endgame", 5)
    booking_process.book_ticket()
